CommandLine/libGMX/gmx.cpp

```diff
@@ -0,0 +1,382 @@
+/** Copyright (C) 2018 Greg Williamson, Robert B. Colton
+***
+*** This file is a part of the ENIGMA Development Environment.
+***
+*** ENIGMA is free software: you can redistribute it and/or modify it under the
+*** terms of the GNU General Public License as published by the Free Software
+*** Foundation, version 3 of the license or any later version.
+***
+*** This application and its source code is distributed AS-IS, WITHOUT ANY
+*** WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#include <pugixml.hpp>
+
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <sstream>
+#include <vector>
+#include <algorithm>
+
+#include "gmx.h"
+
+namespace gmx {
+
+inline std::string string_replace_all(std::string str, std::string substr, std::string nstr)
+{
+  size_t pos = 0;
+  while ((pos = str.find(substr, pos)) != std::string::npos)
+  {
+    str.replace(pos, substr.length(), nstr);
+    pos += nstr.length();
+  }
+  return str;
+}
+
+inline std::string extract_file_path(std::string dir, std::string value) {
+  value = string_replace_all(value, "\\", "/");
+  const std::string end = dir.substr(dir.find_last_of("/") + 1);
+  if (value.find(end) == 0)
+    return dir.substr(0, dir.find_last_of("/") + 1) + value;
+  return dir + "/" + value;
+}
+
+static std::ostream outputStream(nullptr);
+static std::map<std::string, std::vector<std::string>> resMap;
+
+class visited_walker : public pugi::xml_tree_walker {
+  virtual bool for_each(pugi::xml_node &node) {
+    if (node.type() != pugi::node_pcdata  && std::string(node.attribute("visited").value()) != "true")
+      outputStream << "Error: Node at " << node.path() << " was never visited " << std::endl;
+    return true;
+  }
+};
+
+class gmx_root_walker : public pugi::xml_tree_walker {
+  std::map<std::string, int> constants;
+
+  virtual bool for_each(pugi::xml_node &node) {
+    if (node.type() == pugi::node_pcdata) {
+      if (std::string(node.parent().name()) == "constant") {
+        constants[node.parent().attribute("name").value()] = node.text().as_int();
+      } else {
+        resMap[std::string(node.parent().name()) + "s"].push_back(node.value());
+      }
+    }
+    return true;
+  }
+};
+
+std::string FileToString(const std::string &fName) {
```

Comment:
I think we have like five of these elsewhere in the compiler... I guess another doesn't even hurt anything.

Reply:
Yeah same goes for string_replace_all and some other utility methods. We should maybe take another look at this in a future branch to eliminate some of our code redundancy between the compiler and engine. I believe there's also a lot of duplicate file and directory manipulation code.

Reply:
this library will be linked against ide and whatever else so it needs its own string methods

Reply:
Extract the string functions into a shared library... just like you did for protos.

Reply:
thats overkill for 3 line method

Reply:
It's not though, we're also copying string_replace_all around and you're including boost for directory functions Josh already wrote in CompilerSource. Again, extract all this shit, build as shared library.

Reply:
no point. dont need 50 shared libraries. I use boost for dir cause I use it for flags already